File: it.unibo.bls19Local/src/main/java/it/unibo/blsFramework/applLogic/BlsApplicationLogic.java
package it.unibo.blsFramework.applLogic;

import it.unibo.bls.interfaces.ILed;
import it.unibo.bls.utils.Utils;
import it.unibo.blsFramework.interfaces.IAppLogic;
import it.unibo.blsFramework.interfaces.ILedModel;

public class BlsApplicationLogic implements IAppLogic { //Note: it is NOT an observer
	private ILed led;
	private int numOfCalls        = 0;
	private boolean doBlink       = false;
	private final Object monitor  = new Object();
	private boolean stopped       = false;

 	public void setControlled( ILedModel led ){ //no controlled => no activity
		this.led = led;
		doBlinkTheLedWaiting();		//Avoid polling on doBlink (numOfCalls)
	}

	protected void switchTheLed(){
		if( led == null ) return;		//defensive programming
		if( led.getState() ) led.turnOff(); else led.turnOn();
	}

	@Override //from IControlLed
	public void execute( String cmd  ){
		System.out.println("	BlsApplicationLogic | execute cmd=" + cmd);
 		if( cmd.equals("stop") ){ //to allow change in ApplicationLogic
			doBlink = false;
			return;
		}
		numOfCalls++;
		doBlink = numOfCalls % 2 != 0;
		System.out.println("	BlsApplicationLogic | numOfCalls=" + numOfCalls + " doBlink=" + doBlink);
		synchronized( monitor ) {	//see doBlinkTheLedWaiting
 			monitor.notify();
		}
	}

 	//Useful for testing
 	public int getNumOfCalls() {
		return numOfCalls;
	}

	/*
	AVOID POLLING
	 */
	private void doBlinkTheLedWaiting(){
		new Thread(){
			public void run(){
 				System.out.println("	BlsApplicationLogic | doBlinkTheLedWaiting Thread STARTS   ..."  );
				try {
						while ( true  ) {
							synchronized( monitor ) {
								while( ! doBlink ) monitor.wait();
							}
							switchTheLed();
							Utils.delay(250);
						}
  				}catch(Exception e){}
			}
		}.start();
	}
}

File: it.unibo.bls19Local/src/main/kotlin/it/unibo/blsFramework/kotlin/applLogic/BlsApplicationLogic.kt
package it.unibo.blsFramework.kotlin.applLogic

import it.unibo.bls.utils.Utils
import it.unibo.blsFramework.interfaces.IAppLogic
import it.unibo.blsFramework.interfaces.ILedModel
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.channels.SendChannel
import kotlinx.coroutines.channels.actor
import kotlinx.coroutines.launch
import kotlinx.coroutines.newFixedThreadPoolContext

open class BlsApplicationLogic : IAppLogic {
    protected var led: ILedModel? = null
    protected var numCalls = 0
    protected var doBlink = false
    @kotlinx.coroutines.ObsoleteCoroutinesApi
    protected val dispatcher = newFixedThreadPoolContext(2, "mypool")
    //protected val jobBlink = Job();
    /*
    PROACTIVE PART
    The actor returns a send channel
     */
    @kotlinx.coroutines.ObsoleteCoroutinesApi
    protected val actorBlink : SendChannel<String> =
                    GlobalScope.actor<String>(dispatcher, 1 ) {
        for( msg in channel ) {
            println("   ACTOR actorBlink |  msg= $msg doBlink= $doBlink ")
            while ( doBlink ) {
                //switch the led
                if ( led!!.getState() ) led!!.turnOff() else led!!.turnOn()
                //WARNING: inactive for 250 msec
                Utils.delay(250)
            }
            println("   ACTOR actorBlink  | doBlinkTheLed ENDS ...")
        }
    }

    override fun setControlled(led: ILedModel) {
        this.led = led
    }
    //REACTIVE PART
    override fun execute(cmd: String) {
        println("	BlsFrameworkApplicationLogicKt | execute cmd=$cmd  ")
        if (cmd == "stop") { doBlink = false
        }else applLogic( )
    }

    @kotlinx.coroutines.ObsoleteCoroutinesApi
    open  fun applLogic( ){  //open : can be overridden
        numCalls++
        doBlink = numCalls % 2 != 0     //if false actorBlink ends its loop
        println("	BlsFrameworkApplicationLogicKt | execute numCalls=$numCalls doBlink=$doBlink")
        if( doBlink )
            GlobalScope.launch {
                actorBlink.send("work") //REACTIVATES the actor
            }
    }
    //Useful for testing
     override fun getNumOfCalls(): Int {
        return numCalls
    }
}
